### PR TITLE
UI render spans captured with nested rAF callbacks

### PR DESCRIPTION
**Pull request by ssakhamu**
### Implementation

This implementation for UI Render spans relies on the mutation observer API which detects DOM mutation events on the page. 

1.  DOM mutation event is captured and two nested rAF callbacks are created
2. The outer rAF callback executes in the pre-paint phase (before paint is made to the page) and tracks this point as the UI Render start timestamp
3. The inner rAF callback executes in the post-paint phase (after paint is made to the page and before the next paint cycle starts) and tracks this point as the UI Render end timestamp.

After a UI render quiescence period of ~16.67 ms (expected time period between renders), the UI Render span will be created. 

If another mutation occurs within that time, the quiescing UI Render span will be extended.

If an interruption happens on an in-flight UI Render span, the span will end at the time of interruption.
If an interruption happens on a quiescing UI Render span, the span will end at the existing end timestamp for the UI Render.

### Tests

**UI Render**
it('’should create a ui render span that starts at pre-paint raf and ends at post-paint raf”)
it('’should create two ui render spans with two raf cycles separated by the quiescence period')
it(“should create two ui render spans if two raf cycles are separated by the quiescence period but the render quiescence timer callback execution is delayed”)
it (“should create one ui render span if two raf cycles are not separated by the quiescence period”)
it(“should create a single ui render span if there are multiple mutation events at the same time”)
it(“should create a ui render span that ends at the interruption time on interrupt if it has started and not ended”)
it(“should create a ui render span that ends at the ui render completion time on interrupt if it has started and ended but is still quiescing”)

**Root Span**
it(“should end at the interruption time if there are no active network requests and there is an active ui render span”)
it("should end at the last ui render time if there are no active network requests and the ui render span is completed”)

### CODE REVIEW

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 1]
```diff
@@ -15,37 +15,52 @@
...
+   */
+  constructor(span: Span, manager: RootSpanContextManager, tracer: Tracer) {
     this.span = span;
     this.manager = manager;
+    this.tracer = tracer;
 
-    const current = Date.now();
-    this.lastNetworkActivityMs = current;
-    this.lastUiActivityMs = current;
+    this.rootSpanStartAtMs = Date.now();
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: Why is this needed? Can't we get the start timestamp off of the `Span` object?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 2]
```diff
@@ -15,37 +15,52 @@
...
 
+    this.lastNetworkActivityMs = this.rootSpanStartAtMs; // set to current start as default
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: Does this make sense? Should this actually be undefined, given no network requests have happened?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 3]
```diff
@@ -125,15 +145,36 @@ export class RootSpan {
...
-          setTimeout(() => {
-            this.recordUiActivity(); // Refresh quiescence timer due to browser paint
-          }, 0);
+          // capture UI activity start in frame 1
+          const currTimestamp = Date.now();
+
+          if (this.prePaintAtMs === undefined) {
+            this.prePaintAtMs = currTimestamp;
+          } else if (this.postPaintAtMs !== undefined) {
+            const frameDuration = currTimestamp - this.postPaintAtMs;
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: `frameDuration` isn't accurate - this is more like `timeSinceRenderEnd` right?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 4]
```diff
@@ -145,25 +186,101 @@ export class RootSpan {
...
+    }
+  }
+
+  /**
+   * Checks if the UI render span is stable (inactive long enough) and ends it if so.
+   */
+  private endUIRenderSpanIfStable(): void {
+    if (this.interruptedAtMs === undefined) {
+      const timeSinceLastRenderActivity =
+        Date.now() - (this.postPaintAtMs ?? this.rootSpanStartAtMs);
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: `this.rootSpanStartAtMs` is a weird fallback. If `this.postPaintAtMs` is undefined, then we can't end the UI render span, so forcing this variable is weird. I get that it might do the right thing in practice (because it will pass the quiescence check and then call `this.endUiRenderSpan()`, which will do nothing because `this.postPaintAtMs` is undefined), but that's a bit weird/implicit logic.

Why not just add `this.postPaintAtMs !== undefined` to an if check?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 5]
```diff
@@ -145,25 +186,101 @@ export class RootSpan {
...
+      clearTimeout(this.quiescenceRenderTimerId);
+      this.quiescenceRenderTimerId = null;
+    }
+  }
+
+  /**
+   * Checks if the UI render span is stable (inactive long enough) and ends it if so.
+   */
+  private endUIRenderSpanIfStable(): void {
+    if (this.interruptedAtMs === undefined) {
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: I don't understand this - why is this interrupted check here?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 6]
```diff
@@ -145,25 +186,101 @@ export class RootSpan {
...
     }, QUIESCENCE_WINDOW_MS);
   }
 
+  /**
+   * Triggers the closing of the UI render span after a render quiescence window of inactivity.
+   */
+  private renderQuiesce(): void {
+    this.clearRenderQuiesce();
+
+    this.quiescenceRenderTimerId = setTimeout(() => {
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:260]: You kind of have UI render span ending logic in two places, making it hard to understand and adding complexity. Here, in the quiescence window timer, and also in the outer rAF. Why are there both? Do we need both? Can you design a way to make it cleaner?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 7]
```diff
@@ -145,25 +186,101 @@ export class RootSpan {
...
+    }
+  }
+
+  /**
+   * Starts and immediately ends a 'UI Render' span to capture the duration
+   * of the measured UI rendering work.
+   * @param endTimeMs Optional end timestamp. Defaults to the last paint completion time.
+   */
+  private endUIRenderSpan(
+    endTimeMs: number = this.postPaintAtMs ?? this.rootSpanStartAtMs
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: ditto

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 8]
```diff
@@ -145,25 +186,101 @@ export class RootSpan {
...
       this.quiesce(); // Keep waiting, there are active network requests
       return;
     }
 
     const timeSinceLastActivity =
-      Date.now() - Math.max(this.lastNetworkActivityMs, this.lastUiActivityMs);
+      Date.now() -
+      Math.max(
+        this.lastNetworkActivityMs,
+        this.postPaintAtMs ?? this.rootSpanStartAtMs
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: ditto

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 9]
```diff
@@ -172,25 +289,25 @@ export class RootSpan {
...
-    if (this.isInterrupted) {
-      this.span.end(this.lastNetworkActivityMs);
-    } else {
-      this.span.end(
-        Math.max(this.lastNetworkActivityMs, this.lastUiActivityMs)
-      );
-    }
+    const endTime = Math.max(
+      this.lastNetworkActivityMs,
+      this.postPaintAtMs ?? this.interruptedAtMs ?? this.rootSpanStartAtMs
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: ditto

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 10]
```diff
@@ -27,10 +27,19 @@ describe('RootSpanContextManager', () => {
   let sandbox: sinon.SinonSandbox;
   let mockMutationObserver: any;
   let mutationObserverCallback: any;
-  let requestAnimationFrameCallback: any;
+  let queuedAnimationFrameCallbacks: any[] = [];
+
+  const requestAnimationFrameCallbacks = (): void => {
+    const callbacks = [...queuedAnimationFrameCallbacks];
+    queuedAnimationFrameCallbacks.length = 0;
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: why set this array's length to 0 instead of assigning it to a new empty array?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 11]
```diff
@@ -15,37 +15,52 @@
...
-import { Context, Span, type Tracer, trace } from '@opentelemetry/api';
+import { Context, Span, type Tracer, trace, context } from '@opentelemetry/api';
 import { ZoneContextManager } from '@opentelemetry/context-zone';
 
 /**
  * The length of time to wait for activity to cease before closing the root span.
  */
 const QUIESCENCE_WINDOW_MS = 150;
-
+const UI_RENDER_QUIESCENCE_WINDOW_MS = 17; // the time diff between requestAnimationFrame cycles
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: Prefer keeping the same style of comment as above (ie block comments)

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 12]
```diff
@@ -84,24 +102,21 @@ export class RootSpan {
...
+    this.interruptedAtMs = Date.now();
 
     if (this.mutationObserver) {
-      // stop listening for UI mutations
       this.mutationObserver.disconnect();
       this.mutationObserver = null;
     }
 
+    this.endUIRenderSpan(
+      Math.min(this.interruptedAtMs, this.postPaintAtMs ?? this.interruptedAtMs)
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: it's kinda odd to have a scenario where we'd be doing a Math.min between this.interruptedAtMs and this.interruptedAtMs. Could we just do Math.min(this.interruptedAtMs, this.postPaintAtMs)?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 13]
```diff
@@ -27,10 +27,19 @@ describe('RootSpanContextManager', () => {
   let sandbox: sinon.SinonSandbox;
   let mockMutationObserver: any;
   let mutationObserverCallback: any;
-  let requestAnimationFrameCallback: any;
+  let queuedAnimationFrameCallbacks: any[] = [];
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: is this variable used for asserting in tests? if not, can it be removed?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 6 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:260]: it ends in the outer raf because of the off chance that the quiescence timer does expire and the timeout callback is scheduled to run in the event loop, but then a new dom mutation happens and the rendering pipeline triggers. If the previous UI render doesn't close then it will just append the new one to the previous one. The end UI render span logic occurs for this edge case but also the case where no new mutation occurs and timeout callback executes normally.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 14]
```diff
@@ -230,12 +224,66 @@ describe('RootSpanContextManager', () => {
...
+      clock.tick(5); // T = 160
+      requestAnimationFrameCallbacks(); // Frame 1
+
+      clock.tick(5); // T = 165
+      requestAnimationFrameCallbacks(); // Frame 2: postPaintAtMs = 165
 
-      expect(mockSpan.end.calledWith(155)).to.be.true;
+      clock.tick(150); // let quiescence complete (T = 315)
+
+      expect((rootSpan.span as any).end.calledWith(165)).to.be.true;
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: why declare this rootSpan.span as any?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 15]
```diff
@@ -169,41 +178,26 @@ describe('RootSpanContextManager', () => {
       expect(manager.getActiveRootSpan()).to.be.undefined;
     });
 
-    it('should reset timer and set last active timestamp on DOM mutation', () => {
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts:197]: rather than remove this test; can you assert that the opposite is happening? that span does not end on DOM mutation without browser paint?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 16]
```diff
@@ -169,41 +178,26 @@ describe('RootSpanContextManager', () => {
...
       mutationObserverCallback();
 
-      clock.tick(5); // T = 105
-      requestAnimationFrameCallback();
-      clock.tick(100); // T = 205
+      clock.tick(5);
+      requestAnimationFrameCallbacks(); // Frame 1: prePaintAtMs = 105
 
-      expect(mockSpan.end.called).to.be.false;
+      clock.tick(5);
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts:248]: is the clock.tick between rafCallbacks important or can it be removed?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 17]
```diff
@@ -230,12 +224,66 @@ describe('RootSpanContextManager', () => {
...
+      clock.tick(5); // T = 165
+      requestAnimationFrameCallbacks(); // Frame 2: postPaintAtMs = 165
 
-      expect(mockSpan.end.calledWith(155)).to.be.true;
+      clock.tick(150); // let quiescence complete (T = 315)
+
+      expect((rootSpan.span as any).end.calledWith(165)).to.be.true;
+    });
+
+    it('should create a root span that ends at the interruption time if there are no active network requests and an active ui render span', () => {
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: i think the creation of the root span is independent of the test you're asserting here. how about:
it('should end a root span at the interruption time if there are no active network requests and there is an active ui render span') ?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 18]
```diff
@@ -230,12 +224,66 @@ describe('RootSpanContextManager', () => {
...
+
+      // Interrupt at T = 108
+      clock.tick(3);
+      rootSpan.interrupt();
+
+      // Verify that root span was ended with the interruption time (108)
+      expect((rootSpan.span as any).end.calledWith(108)).to.be.true;
+    });
+
+    it('should create a root span that ends at the last ui render time at interruption if there are no active network requests and the ui render span is completed', () => {
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: consider rewording this test name to be clearer

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 19]
```diff
@@ -278,4 +326,266 @@ describe('RootSpanContextManager', () => {
...
+
+      // Run Frame 1 at T = 105
+      clock.tick(5);
+      requestAnimationFrameCallbacks(); // Frame 1: prePaintAtMs = 105
+
+      // Run Frame 2 at T = 110
+      clock.tick(5);
+      requestAnimationFrameCallbacks(); // Frame 2: postPaintAtMs = 110
+
+      // Let UI render quiescence complete (16.67ms)
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: Isn't the value for UI render quiescence 17? consider importing the variable from where it's used and using it here too

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 20]
```diff
@@ -278,4 +326,266 @@ describe('RootSpanContextManager', () => {
...
+
+      // Let everything quiesce (T = 150 + 150 = 300)
+      clock.tick(150);
+
+      // Second UI Render span ended backdated to 150
+      expect(mockSpan).to.not.equal(firstUiSpan);
+      expect(mockSpan.end.calledWith(150)).to.be.true;
+    });
+
+    it('should create two ui render spans if two raf cycles are separated by the quiescence period but the render quiescence timer callback execution is delayed', () => {
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts:582]: this is an interesting test case; I'm interested in learning what led you to adding this one

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts [hunk 21]
```diff
@@ -278,4 +326,266 @@ describe('RootSpanContextManager', () => {
...
+      clock.tick(5);
+      requestAnimationFrameCallbacks(); // Frame 1: prePaintAtMs = 105
+
+      // Run Frame 2 for mutation 1 at T = 110
+      clock.tick(5);
+      requestAnimationFrameCallbacks(); // Frame 2: postPaintAtMs = 110
+
+      // Second mutation at T = 112 (within quiescence window of 110)
+      clock.tick(2);
+      mutationObserverCallback();
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.test.ts:665]: this test case describes two raf cycles separated by quiescence and doesn't mention multiple mutations . Are the mutations necessary for this test to pass?

**Review by MaesterChestnut — COMMENTED** [review]
thanks for providing a thorough list of tests with this change

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 13 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: needed for executing multiple queued callbacks in requestAnimationFrameCallbacks

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 14 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: apparently it needs to be done since .end is a stub that calledWith is called on and the compiler doesn't see that for a normal span type's .end function because it is not a stub.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 20 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:582]: in theory there seems to be a possibility that the render quiescence period could end and the timer callback for creating the UI render span would be in the macrotask queue. However, if a microtask mutation observer callback creates an outer raf callback. That outer raf callback could execute before the macrotask timer callback executes. This is because the rendering pipeline could execute first before the event loop. Therefore, that code path exists in the outer raf callback and a test exists for it.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 21 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:665]: the two individual raf cycles are triggered by the two mutations to the dom, so it would be necessary. requestAnimationFrameCallbacks() would have no callbacks to execute without it.

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 22]
```diff
@@ -125,15 +152,36 @@ export class RootSpan {
...
+            const timeSinceRenderEnd = currTimestamp - this.postPaintAtMs;
+            if (timeSinceRenderEnd >= UI_RENDER_QUIESCENCE_WINDOW_MS) {
+              this.endUIRenderSpan(this.postPaintAtMs);
+              this.prePaintAtMs = currTimestamp;
+            }
+          }
+
+          this.postPaintAtMs = undefined;
+          // quiescence does not exist during a pending postPaint callback
+          this.clearRenderQuiesce();
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:195]: This isn't necessary, because it will already have been called on line 168's `this.endUiRenderSpan()`

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 23]
```diff
@@ -125,15 +152,36 @@ export class RootSpan {
...
+          this.clearQuiesce();
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:196]: Add a comment explaining this line - I'm assuming it's because there's no point in checking every 150ms while we know a render cycle is in progress, right? And in the post rAF callback, we'll explicitly call quiesce again

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 24]
```diff
@@ -145,51 +193,122 @@ export class RootSpan {
...
+    }
+  }
+
+  /**
+   * Starts and immediately ends a 'UI Render' span to capture the duration
+   * of the measured UI rendering work.
+   * @param endTimeMs Optional end timestamp. Defaults to the last paint completion time.
+   */
+  private endUIRenderSpan(
+    endTimeMs: number | undefined = this.postPaintAtMs
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:291]: Instead of having this be passed in, could you just have the method figure out when the end time is (`Math.max(this.postPaintAtMs, this.interruptedAtMs)`)?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 25]
```diff
@@ -145,51 +193,122 @@ export class RootSpan {
...
     // End the span backdated to the exact millisecond work actually stopped
-    if (this.isInterrupted) {
-      this.span.end(this.lastNetworkActivityMs);
+    const activityTimes: number[] = [];
+    if (this.lastNetworkActivityMs !== undefined) {
+      activityTimes.push(this.lastNetworkActivityMs);
+    }
+    if (this.postPaintAtMs !== undefined) {
+      activityTimes.push(this.postPaintAtMs);
+    } else if (this.interruptedAtMs !== undefined) {
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: If `this.postPaintAtMs` is set but is _before_ `this.interruptedAtMs`, we won't add `this.interruptedAtMs` to the array. That's not expected, right?

Comment (rebehe) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts]: What about a simplification like:
```
const activityTimestamps = [this.rootSpanStartAtMs, this.lastNetworkActivityMs, this.postPaintAtMs,  this.interruptedAtMs]
  .filter((t): t is number => t !== undefined);

return Math.max(...validTimes);
```

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 26]
```diff
@@ -145,51 +193,122 @@ export class RootSpan {
...
+    }
+
+    if (activityTimes.length > 0) {
+      const endTime = Math.max(...activityTimes);
+      this.span.end(endTime);
     } else {
-      this.span.end(
-        Math.max(this.lastNetworkActivityMs, this.lastUiActivityMs)
-      );
+      this.span.end(this.rootSpanStartAtMs);
```
Comment (rebehe) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts]: What happens in this case? A 0ms span?

**Review by rebehe — APPROVED** [review]

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 22 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:195]: In the case where a quiescing UI render span gets extended, the endUIRenderSpan won't be called and the existing quiescence would still be ongoing while waiting on the inner raf callback. I suppose it could be moved into an else block on timeSinceRenderEnd >= UI_RENDER_QUIESCENCE_WINDOW_MS, but I would say for readability it would be nice to just see that any outer raf callback results in a temporary halt in quiescence.

Comment (rebehe) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:195]: That's fair 👍

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 21 above)
Comment (MaesterChestnut) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:665]: so by calling mutationObserver, you modify the underlying queuedAnimationFrameCallbacks which is used by requestAnimationFrameCallbacks. Do you call mutationObserverCallback() instead of manually modifying queuedAnimationFrameCallbacks because it more closely resembles what happens when the code runs for real?

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 14 above)
Comment (MaesterChestnut) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: why not stub rootSpan.span instead of pretending it's a stub with the cast to `as any`?

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 24 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:291]: I think it would be cleaner for that flexibility to be kept outside of the method. That way the definition of the function could be read as "now end the ui render span at this chosen time and if there is no chosen time just use what should be the chosen time".

Comment (rebehe) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:291]: I'm ok with that. My only other nit then is to update all callers of `this.endUIRenderSpan` pass in zero arguments, except for the interruption case

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 25 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts]: I like the simplification for including this.rootSpanStartAtMs. The issue with this.interruptedAtMs is that it being greater than this.postPaintAtMs does not mean that it should be the span end time. It should only ever be a span end time if this.postPaintAtMs is undefined. The case is where the UI render span is quiescing and an interruption happens, we wouldn't want the interruption to be the span end time because we want the span end time to backdate to the postPaint time instead. Therefore, there needs to be some way to establish the precedence of this.postPaintAtMs.

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 26 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts]: Yeah it will just be 0ms.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 21 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:665]: I call mutationObserverCallback() to push the callback function for the rAF that is created in the callback function passed into the MutationObserver into queuedAnimationFrameCallbacks.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 16 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:248]: it might not be important, but the two raf callbacks wouldn't execute at the same time so it just avoids any issues there if it were to happen. It is more so to create a realistic scenario.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 14 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts]: As the tests are on RootSpanContextManager, not registering the rootSpan's span through the context manager would probably be an issue. This convention seems to already exist for the tests that exist here so I am just sticking to that.

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 16 above)
Comment (MaesterChestnut) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:248]: consider making a helper function here then to clean up these tests. Something like "doubleRaf(int: timeBetweenRafs)"

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 20 above)
Comment (MaesterChestnut) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:582]: it might be worth adding some of that explanation in as a comment since it's not obvious from this test description when this scenario would appear

File: packages/crashlytics/src/tracing/root-span-context-manager.test.ts (hunk 21 above)
Comment (MaesterChestnut) [reply packages/crashlytics/src/tracing/root-span-context-manager.test.ts:665]: ah it has to do with the actual MutationObserver being called in this code path. thanks for explaining. def a complicated set of logic going on here

**Review by MaesterChestnut — APPROVED** [review]

File: packages/crashlytics/src/tracing/root-span-context-manager.ts [hunk 27]
```diff
@@ -167,68 +213,125 @@ export class RootSpan {
...
+      this.quiescenceRenderTimerId = null;
+    }
+  }
+
+  /**
+   * Starts and immediately ends a 'UI Render' span to capture the duration
+   * of the measured UI rendering work.
+   * @param endTimeMs Optional end timestamp. Defaults to the last paint completion time.
+   */
+  private endUIRenderSpan(
```
Comment (MaesterChestnut) [inline packages/crashlytics/src/tracing/root-span-context-manager.ts:290]: since this  both starts and ends a UI Render span should this be "createUIRenderSpan"?

**Review by MaesterChestnut — APPROVED** [review]

File: packages/crashlytics/src/tracing/root-span-context-manager.ts (hunk 27 above)
Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:290]: I am conflicted on it, the end signifies that the ui render span was in flight and the capturing of it has now ended. However, it is still being created with a start and end time in the end. Though this makes me wonder if the UI render span should be officially started in the outer raf callback instead of in the post hoc creation.

Comment (ssakhamu) [reply packages/crashlytics/src/tracing/root-span-context-manager.ts:290]: will deal with this in a later PR
